src/backend/cuda/kernel/shfl_intrinsics.hpp

```diff
@@ -0,0 +1,112 @@
+/*******************************************************
+ * Copyright (c) 2020, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+namespace cuda {
+namespace kernel {
+
+//__all_sync wrapper
+template<typename T>
+__device__ T all_sync(unsigned mask, T var) {
```

Comment:
Its not always going to be the full mask.

Reply:
Number of instances where full mask was passed to these functions is lot compared to when it isn't in reduce by key implementation itself. It is good to have one less argument when it is possible instead of having the possibility of passing incorrect hard-coded masks all over the code.

Reply:
If you use it that way, it will give you incorrect results in some architectures. You should really think about what you are doing when you use these functions.